internal/kokoro/continuous.sh

```diff
@@ -53,23 +54,87 @@ try3 go mod download
+# but weren't deleted by the current PR.
+# CHANGED_DIRS will be empty when run on master.
+CHANGED_DIRS=$(echo "$SIGNIFICANT_CHANGES" | tr ' ' '\n' | grep "/" | cut -d/ -f1 | sort -u | tr '\n' ' ' | xargs ls -d 2>/dev/null || true)
+
+# List all modules in changed directories.
+# If running on master will collect all modules in the repo, including the root module.
+# shellcheck disable=SC2086
+GO_CHANGED_MODULES="$(find ${CHANGED_DIRS:-.} -name go.mod)"
+# If we didn't find any modules, use the root module.
+GO_CHANGED_MODULES=${GO_CHANGED_MODULES:-./go.mod}
+# Exclude the root module, if present, from the list of sub-modules.
+GO_CHANGED_SUBMODULES=${GO_CHANGED_MODULES#./go.mod}
+
+# Override to determine if all go tests should be run.
+# Does not include static analysis checks.
+RUN_ALL_TESTS="0"
+# If this is a nightly test (not a PR), run all tests.
+if [ -z "${KOKORO_GITHUB_PULL_REQUEST_NUMBER:-}" ]; then
+  RUN_ALL_TESTS="0"
+# If the change touches a repo-spanning file or directory of significance, run all tests.
+elif echo "$SIGNIFICANT_CHANGES" | tr ' ' '\n' | grep "^go.mod$" || [[ $CHANGED_DIRS =~ "internal" ]]; then
+  RUN_ALL_TESTS="1"
+fi
+
+# runTests runs the tests in the current directory. If an argument is specified,
+# it is used as the argument to `go test`.
+runTests() {
+  go test -race -v -timeout 45m "${1:-./...}" 2>&1 \
+    | tee sponge_log.log
+  # Takes the kokoro output log (raw stdout) and creates a machine-parseable
+  # xUnit XML file.
+  cat sponge_log.log \
+    | go-junit-report -set-exit-code > sponge_log.xml
+  # Add the exit codes together so we exit non-zero if any module fails.
+  exit_code=$(($exit_code + $?))
+}
+
 set +e # Run all tests, don't stop after the first failure.
 exit_code=0
-# Run tests and tee output to log file, to be pushed to GCS as artifact.
-for i in `find . -name go.mod`; do
-  pushd `dirname $i`;
-    go test -race -v -timeout 45m ./... 2>&1 \
-      | tee sponge_log.log
-    # Takes the kokoro output log (raw stdout) and creates a machine-parseable
-    # xUnit XML file.
-    cat sponge_log.log \
-      | go-junit-report -set-exit-code > sponge_log.xml
-    # Add the exit codes together so we exit non-zero if any module fails.
-    exit_code=$(($exit_code + $?))
-  popd;
-done
-
-if [[ $KOKORO_BUILD_ARTIFACTS_SUBDIR = *"continuous"* ]]; then
+
+if [[ $RUN_ALL_TESTS = "1" ]]; then
+  echo "Running all tests"
+  # shellcheck disable=SC2044
+  for i in $(find . -name go.mod); do
+    pushd "$(dirname "$i")" > /dev/null;
+      runTests
+    popd > /dev/null;
+  done
+elif [[ -z "${CHANGED_DIRS// }" ]]; then
+  echo "Only running root tests"
+  runTests .
+else
+  runTests . # Always run root tests.
+  echo "Running tests in modified directories: $CHANGED_DIRS"
+  for d in $CHANGED_DIRS; do
+    mods=$(find "$d" -name go.mod)
+    # If there are no modules, just run the tests directly.
+    if [[ -z "$mods" ]]; then
+      pushd "$d" > /dev/null;
+        runTests
```

Comment:
If we have changes in `spanner` and `spanner/spansql` this will run tests for all of spanner twice, right? Do we need to dedupe modules with the changed dirs?

Reply:
I rewrote the script and scrapped this part